### PR TITLE
test/suites: When asserting empty output, get output and test on different lines

### PR DESCRIPTION
**Pull request by markylaing**
We set -e in the test suite to ensure an exit on error. Unfortunately, as far as I can tell -e doesn't apply inside a test. The man page for bash (https://man7.org/linux/man-pages/man1/bash.1.html) has:

> The shell does not exit if the
> command that fails is part of the command list
> immediately following a while or until reserved
> word, part of the test following the if or elif
> reserved words, part of any command executed in a
> && or || list except the command following the
> final && or ||, any command in a pipeline but the
> last (subject to the state of the pipefail shell
> option), or if the command's return value is being
> inverted with !.

Which to me indicates that it should be failing, but at least locally for me it is not.

This means that any test that runs `[ "$(cmd)" = "" ]` is not erroring out it `cmd` fails, and the test is asserting that the stdout from `cmd` is empty. This is usually true if the command fails, as it will send output to stderr instead.

As I see it, the only fix for this is to get the output on one line and test it on the next.

### CODE REVIEW

**Comment by markylaing** [comment]
Lo and behold the tests should never have been passing! https://github.com/canonical/lxd/actions/runs/17823329320/job/50671315261#step:17:17732

**Comment by markylaing** [comment]
Tests are passing now.

**Comment by tomponline** [comment]
> Error: Storage pool driver "ceph" does not support buckets

**Comment by simondeziel** [comment]
> This means that any test that runs `[ "$(cmd)" = "" ]` is not erroring out it `cmd` fails, and the test is asserting that the stdout from `cmd` is empty. This is usually true if the command fails, as it will send output to stderr instead.
> 
> As I see it, the only fix for this is to get the output on one line and test it on the next.

How about this:

```
[ "$(lxc storage bucket list oidc: --all-projects --format csv || echo fail)" = "" ]
```

**Comment by markylaing** [comment]
> > This means that any test that runs `[ "$(cmd)" = "" ]` is not erroring out it `cmd` fails, and the test is asserting that the stdout from `cmd` is empty. This is usually true if the command fails, as it will send output to stderr instead.
> > As I see it, the only fix for this is to get the output on one line and test it on the next.
> 
> How about this:
> 
> ```
> [ "$(lxc storage bucket list oidc: --all-projects --format csv || echo fail)" = "" ]
> ```

That seems like a difficult thing to remember to do. I'd be tempted to start using `lxc query` with `jq` instead as it feels more robust

**Comment by simondeziel** [comment]
> > > This means that any test that runs `[ "$(cmd)" = "" ]` is not erroring out it `cmd` fails, and the test is asserting that the stdout from `cmd` is empty. This is usually true if the command fails, as it will send output to stderr instead.
> > > As I see it, the only fix for this is to get the output on one line and test it on the next.
> > 
> > 
> > How about this:
> > ```
> > [ "$(lxc storage bucket list oidc: --all-projects --format csv || echo fail)" = "" ]
> > ```
> 
> That seems like a difficult thing to remember to do. I'd be tempted to start using `lxc query` with `jq` instead as it feels more robust

I agree on the hard to remember part. I'd be happy to add a linter check that flags all the empty stdout comparison ( ` ...)" = "" ]`) that are missing the `|| echo fail` bit. I prefer the `|| echo fail` alternative to splitting on two lines. WDYT?

Re `lxc query` + `jq` sounds like a good move too.

**Comment by tomponline** [comment]
> Re `lxc query` + `jq` sounds like a good move too.

the downside of that is that it wont help with our CLI code coverage metrics.

**Comment by markylaing** [comment]
> > > > This means that any test that runs `[ "$(cmd)" = "" ]` is not erroring out it `cmd` fails, and the test is asserting that the stdout from `cmd` is empty. This is usually true if the command fails, as it will send output to stderr instead.
> > > > As I see it, the only fix for this is to get the output on one line and test it on the next.
> > > 
> > > 
> > > How about this:
> > > ```
> > > [ "$(lxc storage bucket list oidc: --all-projects --format csv || echo fail)" = "" ]
> > > ```
> > 
> > 
> > That seems like a difficult thing to remember to do. I'd be tempted to start using `lxc query` with `jq` instead as it feels more robust
> 
> I agree on the hard to remember part. I'd be happy to add a linter check that flags all the empty stdout comparison ( ` ...)" = "" ]`) that are missing the `|| echo fail` bit. I prefer the `|| echo fail` alternative to splitting on two lines. WDYT?
> 
> Re `lxc query` + `jq` sounds like a good move too.

Ok happy to go with this and will add a linter to catch future issues

File: test/suites/basic.sh [hunk 1]
```diff
@@ -148,16 +148,13 @@ test_basic_usage() {
...
   # Check volatile.apply_template is altered during rename.
   [ "$(lxc config get bar volatile.apply_template)" = "rename" ]
 
-  [ "$(lxc list -c n | grep -F foo)" = "" ]
-  [ "$(lxc list -c n | grep -F bar)" != "" ]
-
   lxc rename bar foo
-  [ "$(lxc list -c n | grep -F bar)" = "" ]
-  [ "$(lxc list -c n | grep -F foo)" != "" ]
+  [ "$(lxc list -c n -f csv)" = "foo" ]
```
Comment (simondeziel) [inline test/suites/basic.sh:157]: Ah, nice!

File: test/suites/basic.sh [hunk 2]
```diff
@@ -152,11 +152,11 @@ test_basic_usage() {
   # Check volatile.apply_template is altered during rename.
   [ "$(lxc config get bar volatile.apply_template)" = "rename" ]
 
-  [ "$(lxc list -c n | grep -F foo)" = "" ]
+  [ "$(lxc list -c n | grep -F foo || echo fail)" = "" ]
   [ "$(lxc list -c n | grep -F bar)" != "" ]
 
   lxc rename bar foo
-  [ "$(lxc list -c n | grep -F bar)" = "" ]
+  [ "$(lxc list -c n | grep -F bar || echo fail)" = "" ]
```
Comment (simondeziel) [inline test/suites/basic.sh]: I preferred your previous fix of just checking for ` = "foo" ]`

File: test/lint/empty-test-fail.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash -eu
+
+out="$(grep --include \*.sh -rE '^\s+\[\s+"\$\(.*\)"\s+=\s+""\s+]' .)"
```
Comment (simondeziel) [inline test/lint/empty-test-fail.sh]:
```suggestion
out="$(git grep --untracked -E '^\s+\[\s+"\$\(.*\)"\s+=\s+["'\'']{2}\s+]' test/)"
```

This covers both `[ "$(cmd)" = "" ]` and `[ "$(cmd)" = '' ]`.

File: test/lint/empty-test-fail.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash -eu
```
Comment (simondeziel) [inline test/lint/empty-test-fail.sh]: It's best to put the `set -eu` on a separated line to also apply if someone uses `bash test/lint/empty-test-fail.sh` for example.

I'd also add the `set -o pipefail` while at it.

File: test/lint/empty-test-fail.sh (hunk 3 above)
Comment (markylaing) [reply test/lint/empty-test-fail.sh]: Updated the linter, it now uses this:
```
out="$(git grep --untracked -E '^\s+\[\s+"\$\(.*\)"\s+=\s+["'\'']{2}\s+]' test/ | grep -Fv '|| echo fail)' || true)"
```

1. Grep for all `[ $(cmd)" = "" ] ` or `[ "$(cmd)" = '' ].
2. Remove occurrences where `|| echo fail` is used.
3. Don't error if no occurrences are found, or if all occurrences have `|| echo fail`.

File: test/lint/empty-test-fail.sh (hunk 4 above)
Comment (markylaing) [reply test/lint/empty-test-fail.sh]: Done

**Review by tomponline — APPROVED** [review]
Thanks!

**Comment by markylaing** [comment]
@tomponline this is ready when you are, thanks.

**Comment by markylaing** [comment]
> @tomponline this is ready when you are, thanks.

Oh I missed that you'd already approved it :laughing:

**Comment by tomponline** [comment]
> > @tomponline this is ready when you are, thanks.
> 
> Oh I missed that you'd already approved it 😆

yeah i wanted to make sure the tests passed ;)

**Comment by simondeziel** [comment]
Thanks @markylaing that's a really nice improvement!
